src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -49,9 +50,19 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
 
             var stopwatch = SharedStopwatch.StartNew();
 
-            var batchedSourceEntries = sourceTable.Batch();
-            var sourceValues = batchedSourceEntries.SelectAsArray(sourceEntry => sourceEntry.State != EntryState.Removed, sourceEntry => sourceEntry.Item);
-            var sourceInputs = newTable.TrackIncrementalSteps ? batchedSourceEntries.SelectAsArray(sourceEntry => (sourceEntry.Step!, sourceEntry.OutputIndex)) : default;
+            var sourceValuesBuilder = ArrayBuilder<TInput>.GetInstance();
+            var sourceInputsBuilder = newTable.TrackIncrementalSteps ? ArrayBuilder<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>.GetInstance() : null;
+
+            foreach (var entry in sourceTable)
+            {
+                if (entry.State != EntryState.Removed)
+                    sourceValuesBuilder.Add(entry.Item);
+
+                sourceInputsBuilder?.Add((entry.Step!, entry.OutputIndex));
```

Comment:
>sourceInputsBuilder?.Add [](http://example.com/codeflow?start=16&length=24)

Should adding to `sourceInputsBuilder` rely on the same conditions as `sourceValuesBuilder`?

Reply:
no.  we always want to add the information about the steps *including* for a 'remove'.  i can add comments.